flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java

```diff
@@ -969,10 +979,203 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
+        cancelFuture.get();
+        listener.waitForFailed();
+
+        return scheduler.requestJob().getExceptionHistory();
+    }
+
+    @Test
+    public void testExceptionHistoryWithGlobalFailure() throws Exception {
+        final Exception expectedException = new Exception("Expected Global Exception");
+        BiFunction<AdaptiveScheduler, List<ExecutionAttemptID>, Runnable> testLogic =
+                (scheduler, attemptIds) -> {
+                    final ExecutionAttemptID attemptId = attemptIds.remove(0);
+
+                    return () -> {
+                        scheduler.handleGlobalFailure(expectedException);
+                        scheduler.updateTaskExecutionState(
+                                new TaskExecutionStateTransition(
+                                        new TaskExecutionState(
+                                                attemptId, ExecutionState.CANCELED, null)));
+                    };
+                };
+
+        final Iterable<RootExceptionHistoryEntry> actualExceptionHistory =
+                runExceptionHistoryTests(1, testLogic);
+
+        assertThat(actualExceptionHistory).hasSize(1);
+
+        final RootExceptionHistoryEntry failure = actualExceptionHistory.iterator().next();
+        assertThat(failure.getTaskManagerLocation()).isNull();
+        assertThat(failure.getFailingTaskName()).isNull();
+
+        assertThat(failure.getException().deserializeError(classLoader))
+                .isEqualTo(expectedException);
+    }
+
+    @Test
+    public void testExceptionHistoryWithTaskFailure() throws Exception {
+        final Exception expectedException = new Exception("Expected Local Exception");
+        final int numAvailableSlots = 4;
+        BiFunction<AdaptiveScheduler, List<ExecutionAttemptID>, Runnable> testLogic =
+                (scheduler, attemptIds) -> {
+                    final ExecutionAttemptID attemptId = attemptIds.get(1);
+
+                    return () ->
+                            scheduler.updateTaskExecutionState(
+                                    new TaskExecutionStateTransition(
+                                            new TaskExecutionState(
+                                                    attemptId,
+                                                    ExecutionState.FAILED,
+                                                    expectedException)));
+                };
+
+        final Iterable<RootExceptionHistoryEntry> actualExceptionHistory =
+                runExceptionHistoryTests(numAvailableSlots, testLogic);
+
+        assertThat(actualExceptionHistory).hasSize(1);
+
+        final RootExceptionHistoryEntry failure = actualExceptionHistory.iterator().next();
+
+        assertThat(failure.getException().deserializeError(classLoader))
+                .isEqualTo(expectedException);
+    }
+
+    @Test
+    public void testExceptionHistoryWithTaskConcurrentFailure() throws Exception {
```

Comment:
This test fails. But, the reasons run deep. Please let me know, what you think we should do.

**Explanation**
First part: The first task failure 
1. This task is currently is in the execution state `DEPLOYING`
2. It's `Execution` is marked as failed [here](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java#L1122)
3. The `AdaptiveScheduler` transitions to `FAILING`
4. The `DefaultExecutionGraph` calls `failJob`
5. All executions, but the first move to `CANCELING
`
Part two: The second task fails
1. It's already in `CANCELING` execution state
2. It never reaches the part where the failure info is stored in the `Execution` [here](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java#L1122)
3. When we create a `ExceptionHistoryEntry`, we fail [here](https://github.com/metaswirl/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/ExceptionHistoryEntry.java#L59)

Solution proposals
1. We store failure info on an `Execution`, even if it's already canceling (Basically, move the line up that stores the `ErrorInfo` on the `Execution` in `processFail`.
2. We only store a single local failure in the `RootExeptionHistoryEntry`

I propose we go with option two for now. I am not sure what other effects changing the Execution class has.

Reply:
This makes me think that this is actually a bug in `Execution`. On a failure in `CANCELING` state, I'd expect the `Execution` to transition into the `FAILING` state instead of `CANCELED`.

This is something we might want to dig deeper into, but we should do that in another story, so we don't block this effort.

If I understand this correctly, the DefaultScheduler now simply ignores any failures to the `CANCELLING` state (from the exception history perspective).

If the above assumption is correct, I'd suggest we do the same (simply not storing these updates into `failureCollection`) and follow up on this by fixing the Execution state machine (which should IMO fix it for both schedulers).

Hope I didn't miss something obvious, the `Execution` code is bit confusing.

WDYT?

cc @zentol

Reply:
> If I understand this correctly, the DefaultScheduler now simply ignores any failures to the CANCELLING state

Yes, that's also how I read it.

> I'd suggest we do the same

Sounds good.